### PR TITLE
Add support for S3 FileSystem compatible schemes 's3a' and 'oss'

### PR DESCRIPTION
**Pull request by Mionsz**
# Pull-request proposed changes

## AWS S3 compatible implementations (S3Util.h)

Detection made by prefix in file path was extended by two more compatible prefixes
- S3A implementation with `s3a://` prefix added to hive connectors
- OSS implementation with `oss://` prefix added to hive connectors

## Test cases added for S3 filesystems S3A and OSS (S3UtilTest.cpp)

Functional test cases added for proposed vendor specific implementations
of Amazon AWS S3 protocol. Paths and prefixes correctness is being tested.
Future development should register each implementation as independent
filesystem instance for each specified prefix. This is not mandatory now,
as current Velox implementation only implements files reading for S3
based subsystems.

**TODO:** Correctness check for bucket name:
source: https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-s3-bucket-naming-requirements.html
1. Length between 3 and 63
   `3 < length(bucket) < 63`
2. Mandatory label notation regexp:
   `regexp="(^[a-z0-9])([.-]?[a-z0-9]+){2,62}([/]?$)"`
3. Disallowed IPv4 notation:
   `regexp="^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}[/]?$"`

----

# Reference and knowledge base

## Accept S3a and OSS as S3 filesystem

This pull request introduce acceptance of `S3a://` and `oss://` prefixes as supported and compatible `S3` implementations. Alluxio basic implementation and compatibility could also be implemented in an easy way just by sending rest requests to proper API `/api/v1/s3/`

## Difference - s3 vs s3a vs s3n

Naming convention that is being used as prefix for S3 bucket based distributed storage is a big mess this days. Accepting only connections with endpoints identification based on `S3://` protocol prefix only can lead user to incorrect assumptions - that  connection or Velox build is not working correctly. 

Hadoop (Apache) based implementations of AWS S3 protocol started from native block implementation in form of  `s3://` and  `s3n://` which are now deprecated and should not be used and starting from Hadoop 3.x only `s3a://` is supported. 
Amazon on the other hand was using block storage based S3 implementation as `s3bfs://` which is deprecated and now recommended protocol name prefix is `s3://`, so one can not easily distinguish between both implementations only using prefix.

## AWS, Apache, Alibaba

All above formats - as from compatibility and possibility of accessing data implementation - are in most cases interchangeable, it is obvious that data accessing time will be very different. Allowing the use of compatibility protocols like `s3a://` and `oss://` can greatly extend number of people interested in this project.

## Future development consideration - addressing style

Removing prefix from URL may not be enough for some cases, as there are differences between addressing style of `virtual` and `path-style`.

### Hadoop - some older Apache site archives links:

- https://web.archive.org/web/20220531044252/https://cwiki.apache.org/confluence/display/HADOOP2/AmazonS3
- https://web.archive.org/web/20170706121643/https://wiki.apache.org/hadoop/AmazonS3

### AWS Documents:

- https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-plan-file-systems.html#emr-dev-access-file-systems

### OSS compatibility page:

- https://www.alibabacloud.com/help/en/object-storage-service/latest/seamlessly-migrate-data-from-amazon-s3-to-alibaba-cloud-oss
- 
### Addressing styles formats for S3:

- https://documentation.commvault.com/cds/126304_path_style_vs_virtual_hosted_style_requests.html

### CODE REVIEW

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h [hunk 1]
```diff
@@ -57,8 +77,12 @@ inline std::string s3URI(const std::string& bucket, const std::string& key) {
 }
 
 inline std::string s3Path(const std::string_view& path) {
-  // Remove the prefix S3:// from the given path
-  return std::string(path.substr(kS3Scheme.length()));
+  // Remove the prefix 's3://' or one off 'oss://' 's3a://' from the given path
+  if (isS3File(path)) {
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:81]: This should be `isS3AwsFile`?

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h [hunk 2]
```diff
@@ -57,8 +77,12 @@ inline std::string s3URI(const std::string& bucket, const std::string& key) {
...
+    return std::string(path.substr(kS3Scheme.length()));
+  }
+  return std::string(path.substr(
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:84]: let's use explicit checks for each implementation
```
else if (isS3aFile(path)) {
  return std::string(path.substr(kS3aScheme.length()));
} else if (isOssFile(path)) {
  return std::string(path.substr(kOssScheme.length()));
}
return path;
```

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h [hunk 3]
```diff
@@ -30,13 +30,33 @@ namespace facebook::velox {
 
 namespace {
 constexpr std::string_view kSep{"/"};
+// AWS S3 EMRFS, Hadoop block storage filesystem ontop of Amazon S3 buckets
 constexpr std::string_view kS3Scheme{"s3://"};
+// S3N Hadoop S3 native filesystem - s3n:// old implementation (legacy support)
+// This could not be mixed with s3 nor the s3a.
+// S3A Hadoop 3.x, s3n successor - previous connectors "s3" and "s3n" are deprecated.
+constexpr std::string_view kS3aScheme{"s3a://"};
+// OSS Alibaba is another implementation of S3 format, usage for big files only with SSL
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:39]: nit: full stop at the end.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h [hunk 4]
```diff
@@ -57,8 +77,15 @@ inline std::string s3URI(const std::string& bucket, const std::string& key) {
 }
 
 inline std::string s3Path(const std::string_view& path) {
-  // Remove the prefix S3:// from the given path
-  return std::string(path.substr(kS3Scheme.length()));
+  // Remove the prefix 's3://' or one off 'oss://' 's3a://' from the given path
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:80]: nit: full stop at the end.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h [hunk 5]
```diff
@@ -57,8 +77,15 @@ inline std::string s3URI(const std::string& bucket, const std::string& key) {
...
+  if (isS3AwsFile(path)) {
+    return std::string(path.substr(kS3Scheme.length()));
+  } else if (isS3aFile(path)) {
+    return std::string(path.substr(kS3aScheme.length()));
+  } else if (isOssFile(path)) {
+    return std::string(path.substr(kOssScheme.length()));
+  }
+  return path;
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:88]: I guess `std::string(path)`.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h (hunk 4 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/s3fs/S3Util.h:80]: Could simplify write as `Remove one of the prefixes 's3://', 'oss://', 's3a://' if any from the given path.`

**Comment by majetideepak** [comment]
running `make format-fix` should resolve the format check CI test.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h (hunk 4 above)
Comment (Mionsz) [reply velox/connectors/hive/storage_adapters/s3fs/S3Util.h]: fixed

**Comment by Mionsz** [comment]
With this fix I have managed to build and run Prestissimo Worker on our baremetal cluster. Worker is communicating over S3a with MinIO, Hive Metastore is our backend. We are using Presto Benchmark for runs, but there is still this issue for TPC-H (probing tables using SQL works fine): 
image

Any ide how to fix it?

**Comment by majetideepak** [comment]
@Mionsz  see https://github.com/facebookincubator/velox/issues/1340

**Comment by majetideepak** [comment]
@Mionsz Velox also supports both S3 addressing styles. See https://github.com/facebookincubator/velox/blob/5c318cc3360391febb1e91799133fe4a80d3c9a1/velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp#L173-L177

**Comment by majetideepak** [comment]
@Mionsz can you share more details on your benchmarking? Data format, data size, cluster size?

**Comment by Mionsz** [comment]
> @Mionsz can you share more details on your benchmarking? Data format, data size, cluster size?

As for now, we are in development phase so used scale factors are rather small - up to sf300 - for both TPC-H and TPC-DS.
File formats implemented: CSV, ORC, DWRF, Parquet
Cluster size depends on test case, but I can assure that we are able to deploy and run workloads from a scratch, fully automated. If you want a specific info, PM me by e-mail I will forward it to our project manager and give more info then.

**Comment by Mionsz** [comment]
@majetideepak Same here - I have got my corporate CLA signed.

**Comment by majetideepak** [comment]
@Mionsz I somehow missed your update here. Can you also add a couple of tests in `S3UtilTest.cpp` for your changes? Thanks.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h [hunk 6]
```diff
@@ -30,13 +30,32 @@ namespace facebook::velox {
...
+  return filename.substr(0, kS3aScheme.size()) == kS3aScheme;
+}
+
+inline bool isOssFile(const std::string_view filename) {
+  return filename.substr(0, kOssScheme.size()) == kOssScheme;
+}
+
 inline bool isS3File(const std::string_view filename) {
-  return (filename.substr(0, kS3Scheme.size()) == kS3Scheme);
+  // TODO: Each implementation should be registered as separate filesystem
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:57]: nit: each comment must end with a full stop. Probably say
`// TODO: Each prefix should be implemented as its own filesystem.` ?
The registration then is implied.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h [hunk 7]
```diff
@@ -57,8 +76,17 @@ inline std::string s3URI(const std::string& bucket, const std::string& key) {
 }
 
 inline std::string s3Path(const std::string_view& path) {
-  // Remove the prefix S3:// from the given path
-  return std::string(path.substr(kS3Scheme.length()));
+  // Remove one of the prefixes 's3://', 'oss://', 's3a://' if any from the
+  // given path.
+  // TODO: Each implementation should be registered as separate filesystem
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/S3Util.h:81]: `// TODO: Each prefix should be implemented as its own filesystem.` ?

File: velox/connectors/hive/storage_adapters/s3fs/tests/S3UtilTest.cpp [hunk 8]
```diff
@@ -20,16 +20,51 @@
 
 using namespace facebook::velox;
 
+// TODO: Each and every single one prefix S3, S3A, OSS implementation
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/tests/S3UtilTest.cpp:23]: Combine both into
`// TODO: Each prefix should be implemented as its own filesystem.` ?

File: velox/connectors/hive/storage_adapters/s3fs/tests/S3UtilTest.cpp [hunk 9]
```diff
@@ -20,16 +20,51 @@
...
+
+TEST(S3UtilTest, isOssFile) {
+  EXPECT_FALSE(isOssFile("oss:"));
+  EXPECT_FALSE(isOssFile("oss::/bucket"));
+  EXPECT_FALSE(isOssFile("oss:/bucket"));
+  EXPECT_TRUE(isOssFile("OSS://BUCKET/sub-key/file.txt"));
+  EXPECT_TRUE(isOssFile("oss://bucket/file.txt"));
+}
+
+// TODO: Each implementation should be registered as separate filesystem
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/s3fs/tests/S3UtilTest.cpp:60]: `// TODO: Each prefix should be implemented as its own filesystem.` ?

**Review by majetideepak — COMMENTED** [review]
@Mionsz Thanks for adding the tests. I have some comments on the TODOs.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h (hunk 6 above)
Comment (Mionsz) [reply velox/connectors/hive/storage_adapters/s3fs/S3Util.h:57]: True, once again I was in a hurry. Thanks.

File: velox/connectors/hive/storage_adapters/s3fs/S3Util.h (hunk 7 above)
Comment (Mionsz) [reply velox/connectors/hive/storage_adapters/s3fs/S3Util.h:81]: Fixed.

File: velox/connectors/hive/storage_adapters/s3fs/tests/S3UtilTest.cpp (hunk 8 above)
Comment (Mionsz) [reply velox/connectors/hive/storage_adapters/s3fs/tests/S3UtilTest.cpp:23]: Fixed.

**Comment by Mionsz** [comment]
@majetideepak Fixed as you suggested. 
As for the implementation of separate filesystems with registration it should not be a really hard task. Correct me if I am wrong (was not coding in C++17 for long time), but as a starting point one could inherit whole [velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.h](https://github.com/facebookincubator/velox/blob/0ca5f703a8af176589c9b42272c15d0057cee1ec/velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.h#L21) 
```c++
class S3FileSystem : public FileSystem {
```
instantiate each-subclass, than override methods like 
```c++
std::string S3FileSystem::name() const {
```
and add virtual methods to have consistent code: `bucketAndKeyFromS3Path`, `s3Path`, `s3URI`. 
This would be a good starting point for other developers to develop modules like Alluxio (caching engine) etc.

**Comment by Mionsz** [comment]
@majetideepak Now the tests should pass without errors. What is more I have added one more TODO:

**TODO:** Correctness check for bucket name:
source: https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-s3-bucket-naming-requirements.html
1. Length between 3 and 63
   `3 < length(bucket) < 63`
2. Mandatory label notation regexp:
   `regexp="(^[a-z0-9])([.-]?[a-z0-9]+){2,62}([/]?$)"`
3. Disallowed IPv4 notation:
   `regexp="^((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4}[/]?$"`

**Comment by majetideepak** [comment]
@Mionsz thanks for the changes. 
Your comment on the S3 filesystem extension is correct.
The linux-adapters test seems to be failing due to an unrelated error.

**Review by majetideepak — APPROVED** [review]

**Comment by mbasmanova** [comment]
@Mionsz Let me import and land this. Next time, don't hesitate to ping earlier.

**Comment by puchengy** [comment]
> Hadoop (Apache) based implementations of AWS S3 protocol started from native block implementation in form of s3:// and s3n:// which are now deprecated and should not be used and starting from Hadoop 3.x only s3a:// is supported.

@Mionsz any objection if we bring in s3n? There are a lot of hive warehouse configs are still using s3n thus we are not able use Velox

**Pull request closed without merging**